runtime/sema/type_test.go

```diff
@@ -702,3 +702,737 @@ func TestIdentifierCacheUpdate(t *testing.T) {
+			},
+
+			{
+				name: "struct dictionary & resource dictionary",
+				types: []Type{
+					stringStringDictionary,
+					stringResourceDictionary,
+				},
+				expectedSuperType: AnyType,
+			},
+
+			{
+				name: "resource dictionaries",
+				types: []Type{
+					stringResourceDictionary,
+					stringResourceDictionary,
+				},
+				expectedSuperType: stringResourceDictionary,
+			},
+			{
+				name: "resource dictionary & resource",
+				types: []Type{
+					stringResourceDictionary,
+					resourceType,
+				},
+				expectedSuperType: AnyResourceType,
+			},
+			{
+				name: "nested resource dictionaries",
+				types: []Type{
+					nestedResourceDictionary,
+					nestedResourceDictionary,
+				},
+				expectedSuperType: nestedResourceDictionary,
+			},
+			{
+				name: "nested resource-dictionary & nested struct-dictionary",
+				types: []Type{
+					nestedResourceDictionary,
+					nestedStringDictionary,
+				},
+				expectedSuperType: AnyType,
+			},
+		}
+
+		testLeastCommonSuperType(t, tests)
+	})
+
+	t.Run("References types", func(t *testing.T) {
+		t.Parallel()
+
+		tests := []testCase{
+			{
+				name: "homogenous references",
+				types: []Type{
+					&ReferenceType{
+						Type: Int8Type,
+					},
+					&ReferenceType{
+						Type: Int8Type,
+					},
+					&ReferenceType{
+						Type: Int8Type,
+					},
+				},
+				expectedSuperType: &ReferenceType{
+					Type: Int8Type,
+				},
+			},
+			{
+				name: "heterogeneous references",
+				types: []Type{
+					&ReferenceType{
+						Type: Int8Type,
+					},
+					&ReferenceType{
+						Type: StringType,
+					},
+				},
+				expectedSuperType: AnyStructType,
+			},
+			{
+				name: "references & non-references",
+				types: []Type{
+					Int8Type,
+					&ReferenceType{
+						Type: Int8Type,
+					},
+				},
+				expectedSuperType: AnyStructType,
+			},
+			{
+				name: "struct references & resource reference",
+				types: []Type{
+					&ReferenceType{
+						Type: Int8Type,
+					},
+					&ReferenceType{
+						Type: resourceType,
+					},
+				},
+				expectedSuperType: AnyStructType,
+			},
+		}
+
+		testLeastCommonSuperType(t, tests)
+	})
+
+	t.Run("Path types", func(t *testing.T) {
+		t.Parallel()
+
+		tests := []testCase{
+			{
+				name: "homogenous paths",
+				types: []Type{
+					PrivatePathType,
+					PrivatePathType,
+				},
+				expectedSuperType: PrivatePathType,
+			},
+			{
+				name: "capability paths",
+				types: []Type{
+					PrivatePathType,
+					PublicPathType,
+				},
+				expectedSuperType: CapabilityPathType,
+			},
+			{
+				name: "heterogeneous paths",
+				types: []Type{
+					PrivatePathType,
+					PublicPathType,
+					StoragePathType,
+				},
+				expectedSuperType: PathType,
+			},
+			{
+				name: "paths & non-paths",
+				types: []Type{
+					PrivatePathType,
+					PublicPathType,
+					StoragePathType,
+					StringType,
+				},
+				expectedSuperType: AnyStructType,
+			},
+		}
+
+		testLeastCommonSuperType(t, tests)
+	})
+}
```

Comment:
Probably worth adding some test cases for function types, given the complexity of parameter and return type variance.

Reply:
Added in c862949ca8090c03875e9a6bd52ace1a89918203